File: mac/ModelView.m
#import "ModelView.h"
#import "ModelRenderer.h"
#import "MD3PlayerModel.h"
#import "TextureCache.h"
#import <OpenGL/gl3.h>
#import <QuartzCore/CVDisplayLink.h>
#import <math.h>

@implementation ModelView {
    CVDisplayLinkRef _displayLink;
    float _rotationX;
    float _rotationY;
    float _zoom;
    NSPoint _lastMouse;
    BOOL _dragging;
    BOOL _shadersReady;
}

- (instancetype)initWithFrame:(NSRect)frame {
    NSOpenGLPixelFormatAttribute attrs[] = {
        NSOpenGLPFAOpenGLProfile, NSOpenGLProfileVersion3_2Core,
        NSOpenGLPFAColorSize, 24,
        NSOpenGLPFAAlphaSize, 8,
        NSOpenGLPFADepthSize, 24,
        NSOpenGLPFADoubleBuffer,
        NSOpenGLPFAAccelerated,
        NSOpenGLPFASampleBuffers, 1,
        NSOpenGLPFASamples, 4,
        0
    };
    NSOpenGLPixelFormat *pf = [[NSOpenGLPixelFormat alloc] initWithAttributes:attrs];
    self = [super initWithFrame:frame pixelFormat:pf];
    if (self) {
        _rotationX = 0;
        _rotationY = -90;  // Start facing camera
        _zoom = 100.0f;
        _dragging = NO;
        _shadersReady = NO;
        _gamma = 1.0f;
        [self setWantsBestResolutionOpenGLSurface:YES];
    }
    return self;
}

- (void)prepareOpenGL {
    [super prepareOpenGL];

    [[self openGLContext] makeCurrentContext];

    GLint swapInterval = 1;
    [[self openGLContext] setValues:&swapInterval forParameter:NSOpenGLContextParameterSwapInterval];

    glEnable(GL_DEPTH_TEST);
    glEnable(GL_CULL_FACE);
    glFrontFace(GL_CW);  // Q3 winding order
    glCullFace(GL_BACK);
    glEnable(GL_BLEND);
    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
    glClearColor(0.2f, 0.2f, 0.25f, 1.0f);

    if (!_renderer) {
        _renderer = [[ModelRenderer alloc] init];
    }
    _shadersReady = [_renderer setupShaders];
    if (!_shadersReady) {
        NSLog(@"ModelView: shader setup failed");
    }
}

static CVReturn displayLinkCallback(CVDisplayLinkRef displayLink, const CVTimeStamp *now,
                                     const CVTimeStamp *outputTime, CVOptionFlags flagsIn,
                                     CVOptionFlags *flagsOut, void *ctx) {
    @autoreleasepool {
        ModelView *view = (__bridge ModelView *)ctx;
        dispatch_async(dispatch_get_main_queue(), ^{
            [view setNeedsDisplay:YES];
        });
    }
    return kCVReturnSuccess;
}

- (void)startDisplayLink {
    if (_displayLink) return;
    CVDisplayLinkCreateWithActiveCGDisplays(&_displayLink);
    CVDisplayLinkSetOutputCallback(_displayLink, displayLinkCallback, (__bridge void *)self);
    CGLContextObj cglCtx = [[self openGLContext] CGLContextObj];
    CGLPixelFormatObj cglPF = [[self pixelFormat] CGLPixelFormatObj];
    CVDisplayLinkSetCurrentCGDisplayFromOpenGLContext(_displayLink, cglCtx, cglPF);
    CVDisplayLinkStart(_displayLink);
}

- (void)stopDisplayLink {
    if (_displayLink) {
        CVDisplayLinkStop(_displayLink);
        CVDisplayLinkRelease(_displayLink);
        _displayLink = NULL;
    }
}

- (void)dealloc {
    [self stopDisplayLink];
}

- (BOOL)acceptsFirstResponder { return YES; }

- (void)reshape {
    [super reshape];
    [[self openGLContext] makeCurrentContext];
    NSSize size = [self convertSizeToBacking:self.bounds.size];
    glViewport(0, 0, (GLsizei)size.width, (GLsizei)size.height);
}

static void buildPerspective(float *m, float fovY, float aspect, float nearZ, float farZ) {
    float f = 1.0f / tanf(fovY * 0.5f * M_PI / 180.0f);
    memset(m, 0, 16 * sizeof(float));
    m[0] = f / aspect;
    m[5] = f;
    m[10] = (farZ + nearZ) / (nearZ - farZ);
    m[11] = -1.0f;
    m[14] = (2.0f * farZ * nearZ) / (nearZ - farZ);
}

static void buildLookAt(float *m, float eyeX, float eyeY, float eyeZ,
                         float cx, float cy, float cz,
                         float upX, float upY, float upZ) {
    float fx = cx - eyeX, fy = cy - eyeY, fz = cz - eyeZ;
    float flen = sqrtf(fx*fx + fy*fy + fz*fz);
    fx /= flen; fy /= flen; fz /= flen;

    float sx = fy*upZ - fz*upY, sy = fz*upX - fx*upZ, sz = fx*upY - fy*upX;
    float slen = sqrtf(sx*sx + sy*sy + sz*sz);
    sx /= slen; sy /= slen; sz /= slen;

    float ux = sy*fz - sz*fy, uy = sz*fx - sx*fz, uz = sx*fy - sy*fx;

    memset(m, 0, 16 * sizeof(float));
    m[0] = sx;  m[4] = sy;  m[8]  = sz;  m[12] = -(sx*eyeX + sy*eyeY + sz*eyeZ);
    m[1] = ux;  m[5] = uy;  m[9]  = uz;  m[13] = -(ux*eyeX + uy*eyeY + uz*eyeZ);
    m[2] = -fx; m[6] = -fy; m[10] = -fz; m[14] =  (fx*eyeX + fy*eyeY + fz*eyeZ);
    m[3] = 0;   m[7] = 0;   m[11] = 0;   m[15] = 1;
}

- (void)drawRect:(NSRect)dirtyRect {
    [[self openGLContext] makeCurrentContext];

    if (!_shadersReady) return;

    NSSize size = [self convertSizeToBacking:self.bounds.size];
    glViewport(0, 0, (GLsizei)size.width, (GLsizei)size.height);
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

    if (_playerModel && _textureCache) {
        float aspect = size.width / size.height;
        float projMatrix[16];
        buildPerspective(projMatrix, 45.0f, aspect, 1.0f, 2000.0f);

        // Camera orbiting around model center, Z-up
        float centerZ = [_playerModel centerHeight];
        float radX = _rotationX * M_PI / 180.0f;
        float radY = _rotationY * M_PI / 180.0f;
        float camX = _zoom * cosf(radX) * cosf(radY);
        float camY = _zoom * cosf(radX) * sinf(radY);
        float camZ = centerZ + _zoom * sinf(radX);

        float viewMatrix[16];
        buildLookAt(viewMatrix, camX, camY, camZ, 0, 0, centerZ, 0, 0, 1);

        [_playerModel renderWithRenderer:_renderer
                            textureCache:_textureCache
                              viewMatrix:viewMatrix
                              projMatrix:projMatrix
                                   gamma:_gamma];
    }

    [[self openGLContext] flushBuffer];
}

// Mouse interaction
- (void)mouseDown:(NSEvent *)event {
    _dragging = YES;
    _lastMouse = [self convertPoint:[event locationInWindow] fromView:nil];
}

- (void)mouseUp:(NSEvent *)event {
    _dragging = NO;
}

- (void)mouseDragged:(NSEvent *)event {
    NSPoint pos = [self convertPoint:[event locationInWindow] fromView:nil];
    float dx = pos.x - _lastMouse.x;
    float dy = pos.y - _lastMouse.y;
    _rotationY += dx * 0.5f;
    _rotationX += dy * 0.5f;
    // Clamp pitch
    if (_rotationX > 89) _rotationX = 89;
    if (_rotationX < -89) _rotationX = -89;
    _lastMouse = pos;
    [self setNeedsDisplay:YES];
}

- (void)scrollWheel:(NSEvent *)event {
    _zoom -= [event deltaY] * 2.0f;
    if (_zoom < 10) _zoom = 10;
    if (_zoom > 500) _zoom = 500;
    [self setNeedsDisplay:YES];
}

- (void)keyDown:(NSEvent *)event {
    // Arrow keys for frame stepping when paused
    if ([event keyCode] == 123) { // Left arrow
        [_playerModel stepFrame:-1];
        [self setNeedsDisplay:YES];
    } else if ([event keyCode] == 124) { // Right arrow
        [_playerModel stepFrame:1];
        [self setNeedsDisplay:YES];
    } else {
        [super keyDown:event];
    }
}

// Screenshot
- (NSImage *)captureScreenshotWithScale:(int)scale {
    [[self openGLContext] makeCurrentContext];

    NSSize viewSize = [self convertSizeToBacking:self.bounds.size];
    int w = (int)viewSize.width * scale;
    int h = (int)viewSize.height * scale;

    // Create FBO
    GLuint fbo, colorTex, depthRB;
    glGenFramebuffers(1, &fbo);
    glBindFramebuffer(GL_FRAMEBUFFER, fbo);

    glGenTextures(1, &colorTex);
    glBindTexture(GL_TEXTURE_2D, colorTex);
    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, w, h, 0, GL_RGBA, GL_UNSIGNED_BYTE, NULL);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, colorTex, 0);

    glGenRenderbuffers(1, &depthRB);
    glBindRenderbuffer(GL_RENDERBUFFER, depthRB);
    glRenderbufferStorage(GL_RENDERBUFFER, GL_DEPTH_COMPONENT24, w, h);
    glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, depthRB);

    if (glCheckFramebufferStatus(GL_FRAMEBUFFER) != GL_FRAMEBUFFER_COMPLETE) {
        NSLog(@"Screenshot FBO not complete");
        glBindFramebuffer(GL_FRAMEBUFFER, 0);
        glDeleteFramebuffers(1, &fbo);
        glDeleteTextures(1, &colorTex);
        glDeleteRenderbuffers(1, &depthRB);
        return nil;
    }

    glViewport(0, 0, w, h);
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

    if (_playerModel && _textureCache) {
        float aspect = (float)w / (float)h;
        float projMatrix[16];
        buildPerspective(projMatrix, 45.0f, aspect, 1.0f, 2000.0f);

        float centerZ = [_playerModel centerHeight];
        float radX = _rotationX * M_PI / 180.0f;
        float radY = _rotationY * M_PI / 180.0f;
        float camX = _zoom * cosf(radX) * cosf(radY);
        float camY = _zoom * cosf(radX) * sinf(radY);
        float camZ = centerZ + _zoom * sinf(radX);

        float viewMatrix[16];
        buildLookAt(viewMatrix, camX, camY, camZ, 0, 0, centerZ, 0, 0, 1);

        [_playerModel renderWithRenderer:_renderer
                            textureCache:_textureCache
                              viewMatrix:viewMatrix
                              projMatrix:projMatrix
                                   gamma:_gamma];
    }

    // Read pixels as RGB (no alpha)
    unsigned char *pixels = malloc(w * h * 4);
    glReadPixels(0, 0, w, h, GL_RGBA, GL_UNSIGNED_BYTE, pixels);

    // Flip vertically and strip alpha to RGB
    unsigned char *flipped = malloc(w * h * 3);
    for (int row = 0; row < h; row++) {
        unsigned char *src = pixels + (h - 1 - row) * w * 4;
        unsigned char *dst = flipped + row * w * 3;
        for (int x = 0; x < w; x++) {
            dst[x * 3 + 0] = src[x * 4 + 0];
            dst[x * 3 + 1] = src[x * 4 + 1];
            dst[x * 3 + 2] = src[x * 4 + 2];
        }
    }
    free(pixels);

    // Create NSImage (opaque RGB, no alpha)
    NSBitmapImageRep *rep = [[NSBitmapImageRep alloc]
        initWithBitmapDataPlanes:&flipped
                      pixelsWide:w
                      pixelsHigh:h
                   bitsPerSample:8
                 samplesPerPixel:3
                        hasAlpha:NO
                        isPlanar:NO
                  colorSpaceName:NSDeviceRGBColorSpace
                     bytesPerRow:w * 3
                    bitsPerPixel:24];

    NSImage *image = [[NSImage alloc] initWithSize:NSMakeSize(w, h)];
    [image addRepresentation:rep];
    free(flipped);

    // Cleanup
    glBindFramebuffer(GL_FRAMEBUFFER, 0);
    glDeleteFramebuffers(1, &fbo);
    glDeleteTextures(1, &colorTex);
    glDeleteRenderbuffers(1, &depthRB);

    // Restore viewport
    NSSize size = [self convertSizeToBacking:self.bounds.size];
    glViewport(0, 0, (GLsizei)size.width, (GLsizei)size.height);

    return image;
}

- (NSImage *)captureRenderWithWidth:(int)w height:(int)h {
    [[self openGLContext] makeCurrentContext];

    // Create FBO
    GLuint fbo, colorTex, depthRB;
    glGenFramebuffers(1, &fbo);
    glBindFramebuffer(GL_FRAMEBUFFER, fbo);

    glGenTextures(1, &colorTex);
    glBindTexture(GL_TEXTURE_2D, colorTex);
    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, w, h, 0, GL_RGBA, GL_UNSIGNED_BYTE, NULL);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, colorTex, 0);

    glGenRenderbuffers(1, &depthRB);
    glBindRenderbuffer(GL_RENDERBUFFER, depthRB);
    glRenderbufferStorage(GL_RENDERBUFFER, GL_DEPTH_COMPONENT24, w, h);
    glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, depthRB);

    if (glCheckFramebufferStatus(GL_FRAMEBUFFER) != GL_FRAMEBUFFER_COMPLETE) {
        NSLog(@"Render FBO not complete");
        glBindFramebuffer(GL_FRAMEBUFFER, 0);
        glDeleteFramebuffers(1, &fbo);
        glDeleteTextures(1, &colorTex);
        glDeleteRenderbuffers(1, &depthRB);
        return nil;
    }

    glViewport(0, 0, w, h);
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

    if (_playerModel && _textureCache) {
        float renderAspect = (float)w / (float)h;
        float fovY = 45.0f;
        float projMatrix[16];
        buildPerspective(projMatrix, fovY, renderAspect, 1.0f, 2000.0f);

        // Smart framing: compute zoom distance to fit the model's bounding sphere
        // into the render frame, preserving the user's rotation angles.
        float radius = [_playerModel boundingRadius];
        float centerZ = [_playerModel centerHeight];
        float padding = 1.4f; // 40% breathing room

        // Distance needed to fit sphere vertically: r / sin(fov/2)
        float halfFovRad = fovY * 0.5f * M_PI / 180.0f;
        float distV = (radius * padding) / sinf(halfFovRad);
        // Distance needed to fit sphere horizontally: r / sin(hfov/2)
        // where hfov = 2*atan(tan(fov/2)*aspect)
        float halfHFov = atanf(tanf(halfFovRad) * renderAspect);
        float distH = (radius * padding) / sinf(halfHFov);
        // Use the larger (farther) distance so model fits both axes
        float smartZoom = fmaxf(distV, distH);

        float radX = _rotationX * M_PI / 180.0f;
        float radY = _rotationY * M_PI / 180.0f;
        float camX = smartZoom * cosf(radX) * cosf(radY);
        float camY = smartZoom * cosf(radX) * sinf(radY);
        float camZ = centerZ + smartZoom * sinf(radX);

        float viewMatrix[16];
        buildLookAt(viewMatrix, camX, camY, camZ, 0, 0, centerZ, 0, 0, 1);

        [_playerModel renderWithRenderer:_renderer
                            textureCache:_textureCache
                              viewMatrix:viewMatrix
                              projMatrix:projMatrix
                                   gamma:_gamma];
    }

    // Read pixels as RGB (no alpha)
    unsigned char *pixels = malloc(w * h * 4);
    glReadPixels(0, 0, w, h, GL_RGBA, GL_UNSIGNED_BYTE, pixels);

    // Flip vertically and strip alpha to RGB
    unsigned char *flipped = malloc(w * h * 3);
    for (int row = 0; row < h; row++) {
        unsigned char *src = pixels + (h - 1 - row) * w * 4;
        unsigned char *dst = flipped + row * w * 3;
        for (int x = 0; x < w; x++) {
            dst[x * 3 + 0] = src[x * 4 + 0];
            dst[x * 3 + 1] = src[x * 4 + 1];
            dst[x * 3 + 2] = src[x * 4 + 2];
        }
    }
    free(pixels);

    NSBitmapImageRep *rep = [[NSBitmapImageRep alloc]
        initWithBitmapDataPlanes:&flipped
                      pixelsWide:w
                      pixelsHigh:h
                   bitsPerSample:8
                 samplesPerPixel:3
                        hasAlpha:NO
                        isPlanar:NO
                  colorSpaceName:NSDeviceRGBColorSpace
                     bytesPerRow:w * 3
                    bitsPerPixel:24];

    NSImage *image = [[NSImage alloc] initWithSize:NSMakeSize(w, h)];
    [image addRepresentation:rep];
    free(flipped);

    glBindFramebuffer(GL_FRAMEBUFFER, 0);
    glDeleteFramebuffers(1, &fbo);
    glDeleteTextures(1, &colorTex);
    glDeleteRenderbuffers(1, &depthRB);

    NSSize size = [self convertSizeToBacking:self.bounds.size];
    glViewport(0, 0, (GLsizei)size.width, (GLsizei)size.height);

    return image;
}

@end

File: lin/model_view.py
"""GTK4 GLArea-based 3D viewport for MD3 player models."""

import math
import sys

import numpy as np

import gi
gi.require_version('Gtk', '4.0')
from gi.repository import Gtk, Gdk

from OpenGL.GL import *
from PIL import Image

from model_renderer import ModelRenderer


def _build_perspective(fov_y, aspect, near_z, far_z):
    """Build a column-major 4x4 perspective matrix."""
    m = np.zeros(16, dtype=np.float32)
    f = 1.0 / math.tan(fov_y * 0.5 * math.pi / 180.0)
    m[0] = f / aspect
    m[5] = f
    m[10] = (far_z + near_z) / (near_z - far_z)
    m[11] = -1.0
    m[14] = (2.0 * far_z * near_z) / (near_z - far_z)
    return m


def _build_look_at(eye_x, eye_y, eye_z, cx, cy, cz, up_x, up_y, up_z):
    """Build a column-major 4x4 lookAt matrix."""
    fx = cx - eye_x; fy = cy - eye_y; fz = cz - eye_z
    flen = math.sqrt(fx*fx + fy*fy + fz*fz)
    fx /= flen; fy /= flen; fz /= flen

    sx = fy*up_z - fz*up_y; sy = fz*up_x - fx*up_z; sz = fx*up_y - fy*up_x
    slen = math.sqrt(sx*sx + sy*sy + sz*sz)
    sx /= slen; sy /= slen; sz /= slen

    ux = sy*fz - sz*fy; uy = sz*fx - sx*fz; uz = sx*fy - sy*fx

    m = np.zeros(16, dtype=np.float32)
    m[0] = sx;  m[4] = sy;  m[8]  = sz;  m[12] = -(sx*eye_x + sy*eye_y + sz*eye_z)
    m[1] = ux;  m[5] = uy;  m[9]  = uz;  m[13] = -(ux*eye_x + uy*eye_y + uz*eye_z)
    m[2] = -fx; m[6] = -fy; m[10] = -fz; m[14] =  (fx*eye_x + fy*eye_y + fz*eye_z)
    m[3] = 0;   m[7] = 0;   m[11] = 0;   m[15] = 1
    return m


class ModelView(Gtk.GLArea):
    def __init__(self):
        super().__init__()
        self.set_required_version(3, 2)
        self.set_has_depth_buffer(True)
        self.set_auto_render(True)
        self.set_hexpand(True)
        self.set_vexpand(True)

        self.player_model = None
        self.renderer = None
        self.texture_cache = None
        self.gamma = 1.0

        self._rotation_x = 0.0
        self._rotation_y = -90.0  # Start facing camera
        self._zoom = 100.0
        self._shaders_ready = False
        self._drag_start_x = 0.0
        self._drag_start_y = 0.0
        self._tick_callback_id = None

        self.connect('realize', self._on_realize)
        self.connect('unrealize', self._on_unrealize)
        self.connect('render', self._on_render)

        # Mouse drag for orbit
        drag = Gtk.GestureDrag()
        drag.connect('drag-begin', self._on_drag_begin)
        drag.connect('drag-update', self._on_drag_update)
        self.add_controller(drag)

        # Scroll for zoom
        scroll = Gtk.EventControllerScroll(
            flags=Gtk.EventControllerScrollFlags.VERTICAL
        )
        scroll.connect('scroll', self._on_scroll)
        self.add_controller(scroll)

    def _on_realize(self, widget):
        self.make_current()
        if self.get_error() is not None:
            print("GLArea realize error", file=sys.stderr)
            return

        glEnable(GL_DEPTH_TEST)
        glEnable(GL_CULL_FACE)
        glFrontFace(GL_CW)  # Q3 winding order
        glCullFace(GL_BACK)
        glEnable(GL_BLEND)
        glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)
        glClearColor(0.2, 0.2, 0.25, 1.0)

        if self.renderer is None:
            self.renderer = ModelRenderer()
        self._shaders_ready = self.renderer.setup_shaders()
        if not self._shaders_ready:
            print("ModelView: shader setup failed", file=sys.stderr)

        # Start tick callback for continuous rendering
        self._tick_callback_id = self.add_tick_callback(self._tick)

    def _on_unrealize(self, widget):
        self.make_current()
        if self.renderer:
            self.renderer.cleanup()
        if self._tick_callback_id is not None:
            self.remove_tick_callback(self._tick_callback_id)
            self._tick_callback_id = None

    def _tick(self, widget, frame_clock):
        self.queue_render()
        return True  # keep ticking

    def _on_render(self, area, context):
        if not self._shaders_ready:
            return True

        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT)

        if self.player_model and self.texture_cache:
            width = self.get_width()
            height = self.get_height()
            if height == 0:
                height = 1
            # Account for scale factor
            scale = self.get_scale_factor()
            gl_w = width * scale
            gl_h = height * scale
            aspect = gl_w / gl_h

            proj_matrix = _build_perspective(45.0, aspect, 1.0, 2000.0)

            center_z = self.player_model.center_height
            rad_x = self._rotation_x * math.pi / 180.0
            rad_y = self._rotation_y * math.pi / 180.0
            cam_x = self._zoom * math.cos(rad_x) * math.cos(rad_y)
            cam_y = self._zoom * math.cos(rad_x) * math.sin(rad_y)
            cam_z = center_z + self._zoom * math.sin(rad_x)

            view_matrix = _build_look_at(cam_x, cam_y, cam_z, 0, 0, center_z, 0, 0, 1)

            self.player_model.render(self.renderer, self.texture_cache,
                                     view_matrix, proj_matrix, self.gamma)

        return True

    def _on_drag_begin(self, gesture, start_x, start_y):
        self._drag_start_x = start_x
        self._drag_start_y = start_y
        self._drag_last_x = start_x
        self._drag_last_y = start_y

    def _on_drag_update(self, gesture, offset_x, offset_y):
        cur_x = self._drag_start_x + offset_x
        cur_y = self._drag_start_y + offset_y
        dx = cur_x - self._drag_last_x
        dy = cur_y - self._drag_last_y
        self._rotation_y += dx * 0.5
        self._rotation_x -= dy * 0.5  # GTK y-axis is top-down
        if self._rotation_x > 89:
            self._rotation_x = 89
        if self._rotation_x < -89:
            self._rotation_x = -89
        self._drag_last_x = cur_x
        self._drag_last_y = cur_y
        self.queue_render()

    def _on_scroll(self, controller, dx, dy):
        self._zoom += dy * 5.0
        if self._zoom < 10:
            self._zoom = 10
        if self._zoom > 500:
            self._zoom = 500
        self.queue_render()
        return True

    def capture_screenshot(self, scale=2):
        """Capture a screenshot at given scale factor. Returns PIL Image or None."""
        self.make_current()

        width = self.get_width() * self.get_scale_factor() * scale
        height = self.get_height() * self.get_scale_factor() * scale
        w, h = int(width), int(height)

        return self._render_to_image(w, h, use_smart_framing=False)

    def capture_render(self, width, height):
        """Capture a render at specified dimensions with smart framing. Returns PIL Image or None."""
        self.make_current()
        return self._render_to_image(width, height, use_smart_framing=True)

    def _render_to_image(self, w, h, use_smart_framing=False):
        """Internal: render to FBO and return PIL Image."""
        # Create FBO
        fbo = glGenFramebuffers(1)
        glBindFramebuffer(GL_FRAMEBUFFER, fbo)

        color_tex = glGenTextures(1)
        glBindTexture(GL_TEXTURE_2D, color_tex)
        glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, w, h, 0, GL_RGBA, GL_UNSIGNED_BYTE, None)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR)
        glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, color_tex, 0)

        depth_rb = glGenRenderbuffers(1)
        glBindRenderbuffer(GL_RENDERBUFFER, depth_rb)
        glRenderbufferStorage(GL_RENDERBUFFER, GL_DEPTH_COMPONENT24, w, h)
        glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, depth_rb)

        if glCheckFramebufferStatus(GL_FRAMEBUFFER) != GL_FRAMEBUFFER_COMPLETE:
            print("FBO not complete", file=sys.stderr)
            glBindFramebuffer(GL_FRAMEBUFFER, 0)
            glDeleteFramebuffers(1, [fbo])
            glDeleteTextures(1, [color_tex])
            glDeleteRenderbuffers(1, [depth_rb])
            return None

        glViewport(0, 0, w, h)
        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT)

        if self.player_model and self.texture_cache:
            aspect = w / h
            fov_y = 45.0
            proj_matrix = _build_perspective(fov_y, aspect, 1.0, 2000.0)

            center_z = self.player_model.center_height

            if use_smart_framing:
                # Smart framing: compute zoom distance to fit bounding sphere
                radius = self.player_model.bounding_radius
                padding = 1.4
                half_fov_rad = fov_y * 0.5 * math.pi / 180.0
                dist_v = (radius * padding) / math.sin(half_fov_rad)
                half_h_fov = math.atan(math.tan(half_fov_rad) * aspect)
                dist_h = (radius * padding) / math.sin(half_h_fov)
                zoom = max(dist_v, dist_h)
            else:
                zoom = self._zoom

            rad_x = self._rotation_x * math.pi / 180.0
            rad_y = self._rotation_y * math.pi / 180.0
            cam_x = zoom * math.cos(rad_x) * math.cos(rad_y)
            cam_y = zoom * math.cos(rad_x) * math.sin(rad_y)
            cam_z = center_z + zoom * math.sin(rad_x)

            view_matrix = _build_look_at(cam_x, cam_y, cam_z, 0, 0, center_z, 0, 0, 1)

            self.player_model.render(self.renderer, self.texture_cache,
                                     view_matrix, proj_matrix, self.gamma)

        # Read pixels
        pixels = glReadPixels(0, 0, w, h, GL_RGBA, GL_UNSIGNED_BYTE)

        # Cleanup FBO
        glBindFramebuffer(GL_FRAMEBUFFER, 0)
        glDeleteFramebuffers(1, [fbo])
        glDeleteTextures(1, [color_tex])
        glDeleteRenderbuffers(1, [depth_rb])

        # Restore viewport
        scale = self.get_scale_factor()
        glViewport(0, 0, self.get_width() * scale, self.get_height() * scale)

        # Convert to PIL Image (flip vertically, strip alpha)
        img = Image.frombytes('RGBA', (w, h), pixels)
        img = img.transpose(Image.FLIP_TOP_BOTTOM)
        img = img.convert('RGB')
        return img
